Async requests for real!

baronfel opened the pull request:
Fixes #21 

This adds a new static parameter `AsyncInsteadOfTask` and associated plumbing to add new async http calls.

To test, basically every call in the test suite is Async-aware now.
- [x] Fix tests (duplicate parameter names)
- [x] Add in Task/Async switching mechanism

baronfel commented:
Every test failure is from a 'duplicate entry in field table' error, which I suppose means some of the schemas have duplicate fields on typenames?

baronfel commented:
example failing operation definition (note the reused `thingGroupName` field in both the (required) path and (optional) query:
```
{
   "/thing-groups/{thingGroupName}":{
      "delete":{
         "description":"Deletes a thing group.",
         "operationId":"DeleteThingGroup",
         "parameters":[
            {
               "in":"body",
               "name":"body",
               "required":true,
               "schema":{
                  "$ref":"#/definitions/DeleteThingGroupRequest"
               }
            },
            {
               "in":"path",
               "name":"thingGroupName",
               "required":true,
               "type":"string"
            },
            {
               "in":"query",
               "name":"thingGroupName",
               "type":"string"
            }
         ],
         "responses":{
            "200":{
               "description":"Success",
               "schema":{
                  "$ref":"#/definitions/DeleteThingGroupResponse"
               }
            },
            "480":{
               "description":"InvalidRequestException",
               "schema":{
                  "$ref":"#/definitions/InvalidRequestException"
               }
            },
            "481":{
               "description":"VersionConflictException",
               "schema":{
                  "$ref":"#/definitions/VersionConflictException"
               }
            },
            "482":{
               "description":"ThrottlingException",
               "schema":{
                  "$ref":"#/definitions/ThrottlingException"
               }
            },
            "483":{
               "description":"InternalFailureException",
               "schema":{
                  "$ref":"#/definitions/InternalFailureException"
               }
            }
         }
      }
   }
}
```

baronfel commented:
I think we can get around those failures by changing how we generate the parameter names. Instead of a pure mapping across the swagger parameters, we could instead do a fold, accumulating a `string set, ProvidedParameter list` tuple of current parameter names and the provided parameters themselves. Then if we encounter a duplicate parameter name, we append the `parameter.In` to the name to disambiguate. This works because you can't duplicate names across locations.  The only caveat is remembering to reverse the list after the fold so that required parameters are kept before optional ones.

baronfel commented:
@sergey-tihon  how do you feel about each Swagger operation exposing between 1-3 methods?

so for an Operation GetPet, we could generate `GetPet` (sync), `GetPetAsync` (async), and `GetPetTask`(task)?

could we perhaps generate all three as overloads?
`GetPet(string petId) -> Pet`
`GetPet(string petId) -> Async<Pet>`
`GetPet(string petId) -> Task<Pet>`

and let overload resolution do its thing? we could let a user specify an enum as a static param to determine how many of these get generated?

so `type PetStore = SwaggerProvider<..., Methods = Methods.Sync ||| Methods.Async ||| Methods.Task>` to generate all overloads?

baronfel commented:
alright, I fixed the parameter name collisions, and now also made the task/async/sync static parameter that I mentioned above. I started with the same name, different return types strategy because it was the easiest to do.  Then, I had to pin the tests to use async. For that reason I'm considering switching the default to Async. It's 2018, we shouldn't block on Http Requests :D

The test failures from the previous commits are now just github api rate limiting, so if that continues I may mark that test as pending.

sergey-tihon commented:
Thank you for your effort @baronfel! Do not you mid to rebase this PR on top of dev branch? It already contains fixes for parameter duplicates (I did clean up this branch today).

I do like the idea of being able to generate all 3 types of signatures  by user request, but all three by default. I think that we could always generate `GetPet` and `GetPetAsync(..)->Task<Pet>` and provide static parameter `AsyncInsteadOfTask=true`?

For `v1.0` I really want to remove the dependency on `FSharp.Data` `http.fs` implementation and provide a client class that uses `HttpClient` inside, that is why `Task<T>` is a good default. (https://github.com/RSuter/NSwag do similar things)

sergey-tihon commented:
Enum is a nice idea, but I am but sure about `*Task` suffix, I never saw such method names. In the same time `*Async` is the good known naming pattern in dotnet.

baronfel commented:
That all sounds fine to me. I'll rebase the branch and make the changes you suggest.

baronfel commented:
ok, rebased the branch onto dev, added the `Async` suffix where appropriate, and changed the static parameter to toggle the return type of the generated async method

sergey-tihon commented:
> C:\Users\appveyor\AppData\Local\Temp\1\tmp3181.fs (4,28)-(4,120) typecheck error The type provider 'SwaggerProvider.SwaggerTypeProvider' reported an error: The design-time type 'SwaggerProvider.Internal.Compilers.ReflectionHelper' utilized by a type provider was not found in the target reference assembly set '[tgt assembly SwaggerProvider.Runtime, Version=0.10.0.0, Culture=neutral;
 tgt assembly SwaggerProvider, Version=0.10.0.0, Culture=neutral;
 tgt assembly mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;
 tgt assembly FSharp.Core, Version=4.4.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a;
 tgt assembly System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;
 tmp3D59]'. You may be referencing a profile which contains fewer types than those needed by the type provider you are using.

The error means that we emit quotations that use type or methods that not available in runtime, so TP SDK cannot do type rewrite for us. So `ReflectionHelper` and other methods like `cast` that is used in quotations should be in `runtime` assembly

baronfel commented:
That makes perfect sense. One of the things I lost in the rebase was the introduction of that split, I believe. I'll move those functions over, fixup the callsites and check out the test results.

baronfel commented:
Looks like we got there! The appveyor build seems to be hanging on tests for some reason, but they have passed so far.

sergey-tihon commented on src/SwaggerProvider.Runtime/Parser/Schema.fs:
```diff
@@ -145,6 +145,8 @@ type ParameterObject =
       CollectionFormat: CollectionFormat
     }
+    with
+        member x.UnambiguousName = sprintf "%s%A" x.Name x.In
```
Does this method generate nice names? What do you think about
```fsharp
sprintf "%sIn%A" x.Name x.In
```

baronfel replied:
That's a nicer name : thumbsup:

sergey-tihon commented:
@baronfel Very cool! I am happy to merge it, I have just a couple of question.

baronfel commented on .gitignore:
```diff
@@ -179,11 +179,12 @@ release.cmd
 release.sh
 localpackages/
 paket-files
+.paket/load/
```
When you generate-load-scripts with paket they end up in this directory. I use them to quickly test packages in the repl

baronfel commented on src/SwaggerProvider.DesignTime/SwaggerProviderConfig.fs:
```diff
@@ -12,14 +12,23 @@ module private SwaggerProviderConfig =
               ProvidedStaticParameter("IgnoreControllerPrefix", typeof<bool>, true)
-              ProvidedStaticParameter("ProvideNullable", typeof<bool>, false)]
+              ProvidedStaticParameter("ProvideNullable", typeof<bool>, false)
+              ProvidedStaticParameter("AsyncInsteadOfTask", typeof<bool>, false)]
```
I suppose something like `PreferAsyncOverTask` it even just `PreferAsync` would work as well. Names are hard, though :D

sergey-tihon replied:
I trust your choice, go with `PreferAsync` if you like it

baronfel commented:
I think my only reservations remaining with this are that I think that we shouldn't even generate the sync versions of the API calls. It's very little overhead in the user to call . Result on a task or Async.RunSynchronously on an Async, and I think we should encourage non-blocking behaviors

sergey-tihon commented:
Yes, I can accept it... It is 2018, we can do it!
Let's go with `Task` by default and `Async` when `PreferAsync=true`.

What about method names? Will we keep `Async` suffix or not?

baronfel commented:
I'd say drop the Async suffix and let the operation names suffice. If someone has looked at the swagger schema then they'd expect to see those operation names.

baronfel commented:
Brief comment: with `PreferNullable` we default to an F#-friendly type: `Option`. With `PreferAsync` we default to a C#-friendly type. Should we unify one way or the other?

baronfel commented:
- dropped `Async`
- dropped synchronous method generation
- renamed static parameter
- changed `UnambiguousName` generation
- updated docs!

sergey-tihon commented:
Thank you @baronfel !!!

baronfel commented:
Glad to see this done, finally :) Thanks for the review and merge, Sergey!
